text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,342 @@
+| **Authors**     | Bryan Chen, Jiyuan Zheng |
+
+## Summary
+
+This proposal introduces XCQ (Cross Consensus Query), which aims to serve as an intermediary layer between different chain runtime implementations and tools/UIs, to provide a unified interface for cross-chain queries.
+`XCQ` abstracts away concrete implementations across chains and supports custom query computations.
+
+Use cases benefiting from `XCQ` include:
+- XCM bridge UI:
+  - Query asset balances
+  - Query XCM weight and fee from hop and dest chains
+- Wallets:
+  - Query asset balances
+  - Query weights and fees for operations across chains
+- Universal dApp that supports all the parachains:
+  - Perform Feature discovery
+  - Query pallet-specific features
+  - Construct extrinsics by querying pallet index, call index, etc
+
+## Motivation
+
+In Substrate, runtime APIs facilitate off-chain clients in reading the state of the consensus system.
+However, different chains may expose different APIs for a similar query or have varying data types, such as doing custom transformations on direct data, or differing `AccountId` types.
+This diversity also extends to client-side, which may require custom computations over runtime APIs in various use cases.
+Therefore, tools and UI developers often access storage directly and reimplement custom computations to convert data into user-friendly representations, leading to duplicated code between Rust runtime logic and UI JS/TS logic.
+This duplication increases workload and potential for bugs.
+
+Therefore, A system is needed to serve as an intermediary layer between concrete chain runtime implementations and tools/UIs, to provide a unified interface for cross-chain queries.
+
+## Stakeholders
+
+- Runtime Developers
+- Tools/UI Developers
+
+## Explanation
+
+The overall query pattern of the XCQ consists of three components:
+
+- Runtime: View-functions across different pallets are amalgamated through an extension-based system.
+- XCQ query: Custom computations over view-function results are encapsulated via PolkaVM programs.
+- XCQ query arguments: Query arguments like accounts to be queried are also passed together with the query program.
+
+### XCQ Runtime API
+
+The runtime API for off-chain query usage includes two methods:
+
+- `execute_query`: Executes the query and returns the result. It takes the query, input, and weight limit as arguments.
+The query is the query program in PolkaVM program binary format.
+The input is the query arguments that is SCALE-encoded.
+The weight limit is the maximum weight allowed for the query execution.
+- `metadata`: Return metadata of supported extensions (introduced in later section) and methods, serving as a feature discovery functionality
+
+**Example XCQ Runtime API**:
+```rust
+decl_runtime_apis! {
+    pub trait XcqApi {
+        fn execute_query(query: Vec<u8>, input: Vec<u8>, weight_limit: u64) -> XcqResult;
```

Comment:
```suggestion
        fn execute_query(query: Vec<u8>, input: Vec<u8>, gas_limit: u64) -> XcqResult;
```

Is this limit the PVM gas instead of the weight?

Reply:
It seems that the metrics of the PVM gas are in development now, and the gas value is an inner value that should not be exposed to users.